{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/users/lserrano/mweber\n",
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n",
      "/users/lserrano/mweber\n"
     ]
    }
   ],
   "source": [
    "import matplotlib\n",
    "\n",
    "%matplotlib inline\n",
    "    \n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import pandas as pd\n",
    "import json\n",
    "import argparse\n",
    "import re\n",
    "import gzip\n",
    "import urllib\n",
    "from pathlib import Path\n",
    "import textwrap\n",
    "from glob import glob\n",
    "import subprocess\n",
    "import shlex\n",
    "\n",
    "import Bio\n",
    "from Bio import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation, ExactPosition\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "# Use autoreload to keep up-to-date version of custom packages\n",
    "%load_ext autoreload\n",
    "%aimport mwTools\n",
    "%aimport mwTools.bio\n",
    "%aimport cterminal\n",
    "%aimport index_genome_files_bowtie2\n",
    "%autoreload 1\n",
    "from mwTools.paths import p\n",
    "from mwTools.bio import convert_strand_numeric_to_strand_string\n",
    "from mwTools.bio import get_bio_location\n",
    "from mwTools.bio import convert_genbank_to_annotation_df\n",
    "from mwTools.bio import convert_annotation_df_to_bed\n",
    "from cterminal import read_assemblySummary_file, import_genome_gbff_file\n",
    "from index_genome_files_bowtie2 import index_genome_files_bowtie2\n",
    "\n",
    "p = p('isis')\n",
    "print(p.rootPath)\n",
    "refSeqPath = p.refSeqPath\n",
    "rnaSeqDataPath = p.rnaSeqDataPath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "_StoreAction(option_strings=['-d', '--output-dir'], dest='outputDir', nargs=None, const=None, default='.', type=<class 'str'>, choices=None, help='Directory of the indexed genome and BED files.', metavar=None)"
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "argParser = argparse.ArgumentParser()\n",
    "argParser.add_argument('-g', '--genbank', type=str, nargs='+',\n",
    "                       help=\"List of filenames for reference DNA sequences in Genbank format. Wildcards such as ? * and [] can be used.\")\n",
    "argParser.add_argument('-f', '--fasta', type=str, nargs='+',\n",
    "                       help=\"List of filenames for reference DNA sequences in fasta format. Wildcards such as ? * and [] can be used.\")\n",
    "argParser.add_argument('-o', '--output-name', type=str, dest='outputName',\n",
    "                       help=\"Base name of the indexed genome and BED files.\")\n",
    "argParser.add_argument('-d', '--output-dir', type=str, dest='outputDir', default='.',\n",
    "                       help=\"Directory of the indexed genome and BED files.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "args = argParser.parse_args()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "args = argParser.parse_args('-g /users/lserrano/mweber/Databases/RefSeq/Release_2016_10/Genomes_extracted/GCF_000183385.1_ASM18338v1_genomic.gbff /users/lserrano/mweber/Databases/RefSeq/Release_2016_10/Genomes_extracted/GCF_000092585.1_ASM9258v1_genomic.gbff -f /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_013948.1/example.fasta -d /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome'.split())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genbankFileList:\n",
      " /users/lserrano/mweber/Databases/RefSeq/Release_2016_10/Genomes_extracted/GCF_000092585.1_ASM9258v1_genomic.gbff\n",
      "/users/lserrano/mweber/Databases/RefSeq/Release_2016_10/Genomes_extracted/GCF_000183385.1_ASM18338v1_genomic.gbff\n"
     ]
    }
   ],
   "source": [
    "if args.genbank is not None:\n",
    "    genbankFileList = [fn for pattern in args.genbank for fn in glob(pattern) if Path(fn).is_file()]\n",
    "    # Remove duplicated files from list (can happen if multiple patterns match same files)\n",
    "    genbankFileList = sorted(list(set(genbankFileList)))\n",
    "    print(\"genbankFileList:\\n\", \"\\n\".join(genbankFileList))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fastaFiles:\n",
      " /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_013948.1/example.fasta\n"
     ]
    }
   ],
   "source": [
    "if args.fasta is not None:\n",
    "    fastaFileList = [fn for pattern in args.fasta for fn in glob(pattern) if Path(fn).is_file()]\n",
    "    # Remove duplicated files from list (can happen if multiple patterns match same files)\n",
    "    fastaFileList = sorted(list(set(fastaFileList)))\n",
    "    print(\"fastaFiles:\\n\", \"\\n\".join(fastaFileList))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# 2 genome genbank + 2 proteins\n",
    "genbankFileList = ['/users/lserrano/mweber/Databases/RefSeq/Release_2016_10/Genomes_extracted/GCF_000183385.1_ASM18338v1_genomic.gbff',\n",
    "                   '/users/lserrano/mweber/Databases/RefSeq/Release_2016_10/Genomes_extracted/GCF_000092585.1_ASM9258v1_genomic.gbff']\n",
    "fastaFiles = ['/users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_013948.1/example.fasta']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# genome + 2 proteins\n",
    "fastaFiles = ['/users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_013948.1/example.fasta',\n",
    "              '/users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_013948.1/NC_013948.1.fasta']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "outputPath = Path(args.outputDir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mycoplasma_feriruminatoris custom parsing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genbankFileList = ['/users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/Mycoplasma_feriruminatoris/Myfer.annotation.20160921genoma.gbk']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/Mycoplasma_feriruminatoris']"
      ]
     },
     "execution_count": 257,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genbankFileList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fastaFileList = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "outputName = 'Mycoplasma_feriruminatoris'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "outputDir = str(rnaSeqDataPath / 'bowtie2_indexed_genome')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Mycoplasma_feriruminatoris'"
      ]
     },
     "execution_count": 266,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outputName"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/local/lib/python3.5/site-packages/Bio/GenBank/Scanner.py:1230: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       _15myc_circularized_quiver      1075080 bp    DNA     linear   BCT 21-SEP-2016\\n'\n",
      "Found locus '_15myc_circularized_quiver' size '1075080' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  splitline[2], splitline[4]), BiopythonParserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output name: Mycoplasma_feriruminatoris\n",
      "Output path: /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/Mycoplasma_feriruminatoris\n",
      "\n",
      "### Parsing genbank file /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/Mycoplasma_feriruminatoris/Myfer.annotation.20160921genoma.gbk\n",
      "Nb of genome entries in Genbank file: 1\n",
      "Myfer \n",
      "Converting DNA sequence from genbank to fasta.\n",
      "Writing fasta file: /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/Mycoplasma_feriruminatoris/Myfer.fna\n",
      "len(genomeBio.features): 2269\n",
      "CDSDf:\n",
      "   chromosome   end feature  gene         id  locus_tag protein_id  start  \\\n",
      "0      Myfer  1005    gene  dnaA  Myfer_267  Myfer_267       None      0   \n",
      "1      Myfer  1005     CDS  dnaA  Myfer_267  Myfer_267       None      0   \n",
      "2      Myfer  1352    gene  dnaA  Myfer_268  Myfer_268       None   1028   \n",
      "3      Myfer  1352     CDS  dnaA  Myfer_268  Myfer_268       None   1028   \n",
      "4      Myfer  2625    gene  dnaN  Myfer_269  Myfer_269       None   1497   \n",
      "\n",
      "  strand  \n",
      "0      +  \n",
      "1      +  \n",
      "2      +  \n",
      "3      +  \n",
      "4      +  \n",
      "Writing CDS BED file Mycoplasma_feriruminatoris_CDS.bed\n",
      "Writing rRNA BED file Mycoplasma_feriruminatoris_rRNA.bed\n",
      "Writing tRNA BED file Mycoplasma_feriruminatoris_tRNA.bed\n",
      "Writing rRNA_tRNA BED file Mycoplasma_feriruminatoris_rRNA_tRNA.bed\n",
      "We build the bowtie2 index from the following set of fasta files:\n",
      " /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/Mycoplasma_feriruminatoris/Myfer.fna\n",
      "### Running bowtie2-build...\n",
      "/users/lserrano/mweber/Software/bowtie2-2.2.9/bowtie2-build /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/Mycoplasma_feriruminatoris/Myfer.fna Mycoplasma_feriruminatoris\n",
      "b'Settings:\n",
      "  Output files: \"Mycoplasma_feriruminatoris.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/Mycoplasma_feriruminatoris/Myfer.fna\n",
      "Building a SMALL index\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 268770\n",
      "Using parameters --bmax 201578 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 201578 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 2, merged 6; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 1; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 0; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 134384 (target: 201577)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 8\n",
      "  Reserving size (201578) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 104810 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 104811 for bucket 1\n",
      "Getting block 2 of 8\n",
      "  Reserving size (201578) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 175726 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 175727 for bucket 2\n",
      "Getting block 3 of 8\n",
      "  Reserving size (201578) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 169089 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 169090 for bucket 3\n",
      "Getting block 4 of 8\n",
      "  Reserving size (201578) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 115684 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 115685 for bucket 4\n",
      "Getting block 5 of 8\n",
      "  Reserving size (201578) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 102589 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 102590 for bucket 5\n",
      "Getting block 6 of 8\n",
      "  Reserving size (201578) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 190333 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 190334 for bucket 6\n",
      "Getting block 7 of 8\n",
      "  Reserving size (201578) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 143472 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 143473 for bucket 7\n",
      "Getting block 8 of 8\n",
      "  Reserving size (201578) for bucket 8\n",
      "  Calculating Z arrays for bucket 8\n",
      "  Entering block accumulator loop for bucket 8:\n",
      "  bucket 8: 10%\n",
      "  bucket 8: 20%\n",
      "  bucket 8: 30%\n",
      "  bucket 8: 40%\n",
      "  bucket 8: 50%\n",
      "  bucket 8: 60%\n",
      "  bucket 8: 70%\n",
      "  bucket 8: 80%\n",
      "  bucket 8: 90%\n",
      "  bucket 8: 100%\n",
      "  Sorting block of length 73370 for bucket 8\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 73371 for bucket 8\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 404187\n",
      "fchr[G]: 533699\n",
      "fchr[T]: 664953\n",
      "fchr[$]: 1075080\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4552872 bytes to primary EBWT file: Mycoplasma_feriruminatoris.1.bt2\n",
      "Wrote 268776 bytes to secondary EBWT file: Mycoplasma_feriruminatoris.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 1075080\n",
      "    bwtLen: 1075081\n",
      "    sz: 268770\n",
      "    bwtSz: 268771\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 67193\n",
      "    offsSz: 268772\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 5600\n",
      "    numLines: 5600\n",
      "    ebwtTotLen: 358400\n",
      "    ebwtTotSz: 358400\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:01\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 268770\n",
      "Using parameters --bmax 201578 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 201578 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 6; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 1; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 153582 (target: 201577)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (201578) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 191202 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 191203 for bucket 1\n",
      "Getting block 2 of 7\n",
      "  Reserving size (201578) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 169701 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 169702 for bucket 2\n",
      "Getting block 3 of 7\n",
      "  Reserving size (201578) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 155431 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 155432 for bucket 3\n",
      "Getting block 4 of 7\n",
      "  Reserving size (201578) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 149530 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 149531 for bucket 4\n",
      "Getting block 5 of 7\n",
      "  Reserving size (201578) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 121575 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 121576 for bucket 5\n",
      "Getting block 6 of 7\n",
      "  Reserving size (201578) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 194867 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 194868 for bucket 6\n",
      "Getting block 7 of 7\n",
      "  Reserving size (201578) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 92768 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 92769 for bucket 7\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 404187\n",
      "fchr[G]: 533699\n",
      "fchr[T]: 664953\n",
      "fchr[$]: 1075080\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4552872 bytes to primary EBWT file: Mycoplasma_feriruminatoris.rev.1.bt2\n",
      "Wrote 268776 bytes to secondary EBWT file: Mycoplasma_feriruminatoris.rev.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 1075080\n",
      "    bwtLen: 1075081\n",
      "    sz: 268770\n",
      "    bwtSz: 268771\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 67193\n",
      "    offsSz: 268772\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 5600\n",
      "    numLines: 5600\n",
      "    ebwtTotLen: 358400\n",
      "    ebwtTotSz: 358400\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:02\n",
      "'\n",
      "### bowtie2-build finished\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'/users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/Mycoplasma_feriruminatoris/Mycoplasma_feriruminatoris'"
      ]
     },
     "execution_count": 268,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indexedGenomePath = index_genome_files_bowtie2(genbankFileList=genbankFileList, fastaFileList=fastaFileList,\n",
    "                                             outputName=outputName, outputDir=outputDir)\n",
    "indexedGenomePath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/Mycoplasma_feriruminatoris/mferiruminatoris_small_aa.fa')"
      ]
     },
     "execution_count": 256,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnaSeqDataPath / 'bowtie2_indexed_genome' / 'Mycoplasma_feriruminatoris' / 'mferiruminatoris_small_aa.fa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 275,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERROR: mf41120 |-|53..-24\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>end</th>\n",
       "      <th>feature</th>\n",
       "      <th>id</th>\n",
       "      <th>start</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>523404</td>\n",
       "      <td>CDS</td>\n",
       "      <td>mf20623</td>\n",
       "      <td>523404</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>146932</td>\n",
       "      <td>CDS</td>\n",
       "      <td>mf48351</td>\n",
       "      <td>146932</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>196742</td>\n",
       "      <td>CDS</td>\n",
       "      <td>mf09589</td>\n",
       "      <td>196742</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>361628</td>\n",
       "      <td>CDS</td>\n",
       "      <td>mf10636</td>\n",
       "      <td>361628</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>64116</td>\n",
       "      <td>CDS</td>\n",
       "      <td>mf17036</td>\n",
       "      <td>64116</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      end feature       id   start strand\n",
       "0  523404     CDS  mf20623  523404      +\n",
       "1  146932     CDS  mf48351  146932      -\n",
       "2  196742     CDS  mf09589  196742      +\n",
       "3  361628     CDS  mf10636  361628      +\n",
       "4   64116     CDS  mf17036   64116      +"
      ]
     },
     "execution_count": 275,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotList = []\n",
    "for record in SeqIO.parse(str(rnaSeqDataPath / 'bowtie2_indexed_genome' / 'Mycoplasma_feriruminatoris' / 'mferiruminatoris_small_aa.fa'), format=\"fasta\"):\n",
    "    regexMatch = re.search(r'(mf\\d+)\\s*\\|([+-])\\|(\\d+)\\.\\.(\\d+)', record.description)\n",
    "#     recordId = regexMatch.group(1)\n",
    "#     recordStrand = regexMatch.group(2)\n",
    "#     recordStart = int(regexMatch.group(3))\n",
    "#     recordEnd = int(regexMatch.group(4))\n",
    "#     print(record.description, recordId, recordStrand, recordStart, recordEnd)\n",
    "    if regexMatch:\n",
    "        annotList.append({\n",
    "                            'feature':'CDS',\n",
    "                            'id':regexMatch.group(1),\n",
    "                            'strand':regexMatch.group(2),\n",
    "                            'start':int(regexMatch.group(3)),\n",
    "                            'end':int(regexMatch.group(3))\n",
    "                         })\n",
    "    else:\n",
    "        print(\"ERROR:\", record.description)\n",
    "\n",
    "annotDf = pd.DataFrame(annotList)\n",
    "annotDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "convert_annotation_df_to_bed(CDSDf)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parse Genbank files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output name: NC_004829.2\n",
      "Output path: /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_004829.2/NC_004829.2\n",
      "\n",
      "### Parsing genbank file /users/lserrano/mweber/Databases/RefSeq/Release_2016_10/Genomes_extracted/GCF_000092585.1_ASM9258v1_genomic.gbff\n",
      "Nb of genome entries in Genbank file: 1\n",
      "NC_004829.2 Mycoplasma gallisepticum str. R(low), complete genome.\n",
      "Converting DNA sequence from genbank to fasta.\n",
      "Writing fasta file: /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_004829.2/NC_004829.2/NC_004829.2.fna\n",
      "len(genomeBio.features): 1761\n",
      "CDSDf:\n",
      "     chromosome   end feature  gene           id    locus_tag      protein_id  \\\n",
      "0  NC_004829.2  2681    gene  None  MGA_RS00005  MGA_RS00005            None   \n",
      "1  NC_004829.2  2681     CDS  None  MGA_RS00005  MGA_RS00005  WP_011113230.1   \n",
      "2  NC_004829.2  3147    gene  None  MGA_RS00010  MGA_RS00010            None   \n",
      "3  NC_004829.2  3147     CDS  None  MGA_RS00010  MGA_RS00010  WP_011113231.1   \n",
      "4  NC_004829.2  4548    gene  None  MGA_RS00015  MGA_RS00015            None   \n",
      "\n",
      "   start strand  \n",
      "0   1892      +  \n",
      "1   1892      +  \n",
      "2   2670      +  \n",
      "3   2670      +  \n",
      "4   3162      +  \n",
      "\n",
      "### Parsing genbank file /users/lserrano/mweber/Databases/RefSeq/Release_2016_10/Genomes_extracted/GCF_000183385.1_ASM18338v1_genomic.gbff\n",
      "Nb of genome entries in Genbank file: 1\n",
      "NC_014760.1 Mycoplasma bovis PG45 clone MU clone A2, complete genome.\n",
      "Converting DNA sequence from genbank to fasta.\n",
      "Writing fasta file: /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_004829.2/NC_004829.2/NC_014760.1.fna\n",
      "len(genomeBio.features): 1924\n",
      "CDSDf:\n",
      "     chromosome   end feature  gene                id         locus_tag  \\\n",
      "0  NC_014760.1  1409    gene  None  MBOVPG45_RS00005  MBOVPG45_RS00005   \n",
      "1  NC_014760.1  1409     CDS  None  MBOVPG45_RS00005  MBOVPG45_RS00005   \n",
      "2  NC_014760.1  2652    gene  None  MBOVPG45_RS00010  MBOVPG45_RS00010   \n",
      "3  NC_014760.1  2652     CDS  None  MBOVPG45_RS00010  MBOVPG45_RS00010   \n",
      "4  NC_014760.1  2882    gene  None  MBOVPG45_RS00015  MBOVPG45_RS00015   \n",
      "\n",
      "       protein_id  start strand  \n",
      "0            None      8      +  \n",
      "1  WP_013456605.1      8      +  \n",
      "2            None   1542      +  \n",
      "3  WP_013456452.1   1542      +  \n",
      "4            None   2654      +  \n"
     ]
    }
   ],
   "source": [
    "CDSDfList = []\n",
    "fastaFileToBeIndexedList = []\n",
    "genomeList = []\n",
    "\n",
    "if genbankFileList is not None:\n",
    "    \n",
    "    # Import the first file to get the genome id\n",
    "    \n",
    "    genomeIdFirst = next(Bio.SeqIO.parse(genbankFileList[0], format=\"genbank\")).id\n",
    "\n",
    "    if args.outputName is None:\n",
    "        if genomeIdFirst is not None:\n",
    "            outputName = genomeIdFirst\n",
    "        else:\n",
    "            print(\"ERROR: genome id could not be extracted from the first genbank file,\", genbankFileList[0], \"please use the -o option to set the output name.\")\n",
    "    else:\n",
    "        outputName = args.outputName\n",
    "    print(\"Output name:\", outputName)\n",
    "    \n",
    "    outputPath = outputPath / outputName\n",
    "    outputPath.mkdir(exist_ok=True)\n",
    "    print(\"Output path:\", outputPath)\n",
    "    \n",
    "    \n",
    "    for genbankFilename in genbankFileList:\n",
    "\n",
    "        genbankPath = Path(genbankFilename).parent\n",
    "\n",
    "        genbankBioList = list(Bio.SeqIO.parse(genbankFilename, format=\"genbank\"))\n",
    "        print(\"\\n### Parsing genbank file\", genbankFilename)\n",
    "        print(\"Nb of genome entries in Genbank file:\", len(genbankBioList))\n",
    "\n",
    "        for genomeBio in genbankBioList:\n",
    "            \n",
    "            print(genomeBio.id, genomeBio.description)\n",
    "            \n",
    "            # Convert genome DNA sequence to FASTA nucleotide.\n",
    "            # Test if the genbank has a well-defined DNA sequence.\n",
    "            hasWellDefinedSeq = True\n",
    "            hasWellDefinedSeq = \\\n",
    "                not (\n",
    "                (genomeBio.seq.count('N') / len(genomeBio.seq)) > 0.5\n",
    "                or genomeBio.seq == ''\n",
    "                )\n",
    "\n",
    "            if hasWellDefinedSeq:\n",
    "                print(\"Converting DNA sequence from genbank to fasta.\")\n",
    "                fastaId = genomeBio.id\n",
    "                fastaDescription = genomeBio.description\n",
    "                genomeSeq = textwrap.fill(str(genomeBio.seq), width=70)\n",
    "                genomeFastaFile = str(outputPath / (fastaId + '.fna'))\n",
    "                fastaFileToBeIndexedList.append(genomeFastaFile)\n",
    "\n",
    "                with open(genomeFastaFile, 'w') as f:\n",
    "                    f.write(\">{}|{}\\n{}\".format(fastaId, fastaDescription, genomeSeq))\n",
    "                print(\"Writing fasta file:\", genomeFastaFile)\n",
    "                \n",
    "                genomeList.append({'chromosome':fastaId, 'length':len(genomeBio.seq)})\n",
    "\n",
    "            # Convert genbank annotations to BED file\n",
    "            CDSDf = convert_genbank_to_annotation_df(genomeBio, verbose=1)\n",
    "            CDSDfList.append(CDSDf)\n",
    "            print(\"CDSDf:\\n\", CDSDf.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing CDS BED file NC_004829.2_CDS.bed\n",
      "Writing rRNA BED file NC_004829.2_rRNA.bed\n",
      "Writing tRNA BED file NC_004829.2_tRNA.bed\n",
      "Writing rRNA_tRNA BED file NC_004829.2_rRNA_tRNA.bed\n"
     ]
    }
   ],
   "source": [
    "if CDSDfList != []:\n",
    "\n",
    "    CDSDf = pd.concat(CDSDfList)\n",
    "\n",
    "    with (outputPath / (outputName + '_CDS.bed')).open('w') as f:\n",
    "        f.write(convert_annotation_df_to_bed(CDSDf))\n",
    "    print(\"Writing CDS BED file\", (outputName + '_CDS.bed'))\n",
    "\n",
    "    rRNADf = CDSDf[CDSDf['feature'] == 'rRNA']\n",
    "    rRNADf\n",
    "    with (outputPath / (outputName + '_rRNA.bed')).open('w') as f:\n",
    "        f.write(convert_annotation_df_to_bed(rRNADf))\n",
    "    print(\"Writing rRNA BED file\", (outputName + '_rRNA.bed'))\n",
    "\n",
    "    tRNADf = CDSDf[CDSDf['feature'] == 'tRNA']\n",
    "    tRNADf.head()\n",
    "    with (outputPath / (outputName + '_tRNA.bed')).open('w') as f:\n",
    "        f.write(convert_annotation_df_to_bed(tRNADf))\n",
    "    print(\"Writing tRNA BED file\", (outputName + '_tRNA.bed'))\n",
    "\n",
    "    with (outputPath / (outputName + '_rRNA_tRNA.bed')).open('w') as f:\n",
    "        f.write(convert_annotation_df_to_bed(pd.concat([tRNADf, rRNADf])))\n",
    "    print(\"Writing rRNA_tRNA BED file\", (outputName + '_rRNA_tRNA.bed'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import fasta files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if fastaFileList is not None:\n",
    "    fastaFileToBeIndexedList += fastaFileList\n",
    "\n",
    "for fastaFile in fastaFileList:\n",
    "    for fastaBio in SeqIO.parse(fastaFile, format=\"fasta\"):\n",
    "        genomeList.append({'chromosome':fastaBio.id, 'length':len(fastaBio.seq)})\n",
    "\n",
    "with (outputPath / (outputName + '_genome.bed')).open('w') as f:\n",
    "    for genome in genomeList:\n",
    "        f.write(\"{}\\t{}\\n\".format(genome['chromosome'], genome['length']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Index reference genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We build the bowtie2 index from the following set of fasta files:\n",
      " /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_004829.2/NC_004829.2.fna\n",
      "/users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_004829.2/NC_014760.1.fna\n",
      "/users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_013948.1/example.fasta\n"
     ]
    }
   ],
   "source": [
    "print(\"We build the bowtie2 index from the following set of fasta files:\\n\", \"\\n\".join(fastaFileToBeIndexedList))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "bowtie2Build = '/users/lserrano/mweber/Software/bowtie2-2.2.9/bowtie2-build'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### Running bowtie2-build...\n",
      "/users/lserrano/mweber/Software/bowtie2-2.2.9/bowtie2-build /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_004829.2/NC_004829.2.fna,/users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_004829.2/NC_014760.1.fna,/users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_013948.1/example.fasta NC_004829.2\n"
     ]
    }
   ],
   "source": [
    "print(\"### Running bowtie2-build...\")\n",
    "cmd = bowtie2Build + ' ' + \",\".join(fastaFileToBeIndexedList) + \" \" + outputName\n",
    "print(cmd)\n",
    "cmd = shlex.split(cmd)\n",
    "cmd_output = subprocess.check_output(cmd, stderr=subprocess.STDOUT)\n",
    "cmd_output = re.sub(r'\\\\n','\\n',str(cmd_output))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "b'Settings:\n",
      "  Output files: \"NC_004829.2.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_004829.2/NC_004829.2.fna\n",
      "  /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_004829.2/NC_014760.1.fna\n",
      "  /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_013948.1/example.fasta\n",
      "Building a SMALL index\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 504201\n",
      "Using parameters --bmax 378151 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 378151 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 6; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 1; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 252100 (target: 378150)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 8\n",
      "  Reserving size (378151) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 172435 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 172436 for bucket 1\n",
      "Getting block 2 of 8\n",
      "  Reserving size (378151) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 374225 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 374226 for bucket 2\n",
      "Getting block 3 of 8\n",
      "  Reserving size (378151) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 152520 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 152521 for bucket 3\n",
      "Getting block 4 of 8\n",
      "  Reserving size (378151) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 280245 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 280246 for bucket 4\n",
      "Getting block 5 of 8\n",
      "  Reserving size (378151) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 328997 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 328998 for bucket 5\n",
      "Getting block 6 of 8\n",
      "  Reserving size (378151) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 158853 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 158854 for bucket 6\n",
      "Getting block 7 of 8\n",
      "  Reserving size (378151) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 272405 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 272406 for bucket 7\n",
      "Getting block 8 of 8\n",
      "  Reserving size (378151) for bucket 8\n",
      "  Calculating Z arrays for bucket 8\n",
      "  Entering block accumulator loop for bucket 8:\n",
      "  bucket 8: 10%\n",
      "  bucket 8: 20%\n",
      "  bucket 8: 30%\n",
      "  bucket 8: 40%\n",
      "  bucket 8: 50%\n",
      "  bucket 8: 60%\n",
      "  bucket 8: 70%\n",
      "  bucket 8: 80%\n",
      "  bucket 8: 90%\n",
      "  bucket 8: 100%\n",
      "  Sorting block of length 277117 for bucket 8\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 277118 for bucket 8\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 700717\n",
      "fchr[G]: 1008889\n",
      "fchr[T]: 1313758\n",
      "fchr[$]: 2016804\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4866992 bytes to primary EBWT file: NC_004829.2.1.bt2\n",
      "Wrote 504208 bytes to secondary EBWT file: NC_004829.2.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 2016804\n",
      "    bwtLen: 2016805\n",
      "    sz: 504201\n",
      "    bwtSz: 504202\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 126051\n",
      "    offsSz: 504204\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 10505\n",
      "    numLines: 10505\n",
      "    ebwtTotLen: 672320\n",
      "    ebwtTotSz: 672320\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:02\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 504201\n",
      "Using parameters --bmax 378151 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 378151 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 6; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 1; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 288114 (target: 378150)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (378151) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 343375 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 343376 for bucket 1\n",
      "Getting block 2 of 7\n",
      "  Reserving size (378151) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 273058 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 273059 for bucket 2\n",
      "Getting block 3 of 7\n",
      "  Reserving size (378151) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 277405 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 277406 for bucket 3\n",
      "Getting block 4 of 7\n",
      "  Reserving size (378151) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 317501 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 317502 for bucket 4\n",
      "Getting block 5 of 7\n",
      "  Reserving size (378151) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 194472 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 194473 for bucket 5\n",
      "Getting block 6 of 7\n",
      "  Reserving size (378151) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 369618 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 369619 for bucket 6\n",
      "Getting block 7 of 7\n",
      "  Reserving size (378151) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 241369 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 241370 for bucket 7\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 700717\n",
      "fchr[G]: 1008889\n",
      "fchr[T]: 1313758\n",
      "fchr[$]: 2016804\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4866992 bytes to primary EBWT file: NC_004829.2.rev.1.bt2\n",
      "Wrote 504208 bytes to secondary EBWT file: NC_004829.2.rev.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 2016804\n",
      "    bwtLen: 2016805\n",
      "    sz: 504201\n",
      "    bwtSz: 504202\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 126051\n",
      "    offsSz: 504204\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 10505\n",
      "    numLines: 10505\n",
      "    ebwtTotLen: 672320\n",
      "    ebwtTotSz: 672320\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:02\n",
      "'\n",
      "### bowtie2-build finished\n"
     ]
    }
   ],
   "source": [
    "print(cmd_output)\n",
    "print(\"### bowtie2-build finished\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Testing script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/users/lserrano/mweber/Databases/RefSeq/Release_2016_10/Genomes_extracted/GCF_000092585.1_ASM9258v1_genomic.gbff',\n",
       " '/users/lserrano/mweber/Databases/RefSeq/Release_2016_10/Genomes_extracted/GCF_000183385.1_ASM18338v1_genomic.gbff']"
      ]
     },
     "execution_count": 222,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genbankFileList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_013948.1/example.fasta']"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fastaFileList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "args.outputName"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome'"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "args.outputDir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output name: NC_004829.2\n",
      "Output path: /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_004829.2\n",
      "\n",
      "### Parsing genbank file /users/lserrano/mweber/Databases/RefSeq/Release_2016_10/Genomes_extracted/GCF_000092585.1_ASM9258v1_genomic.gbff\n",
      "Nb of genome entries in Genbank file: 1\n",
      "NC_004829.2 Mycoplasma gallisepticum str. R(low), complete genome.\n",
      "Converting DNA sequence from genbank to fasta.\n",
      "Writing fasta file: /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_004829.2/NC_004829.2.fna\n",
      "len(genomeBio.features): 1761\n",
      "CDSDf:\n",
      "     chromosome   end feature  gene           id    locus_tag      protein_id  \\\n",
      "0  NC_004829.2  2681    gene  None  MGA_RS00005  MGA_RS00005            None   \n",
      "1  NC_004829.2  2681     CDS  None  MGA_RS00005  MGA_RS00005  WP_011113230.1   \n",
      "2  NC_004829.2  3147    gene  None  MGA_RS00010  MGA_RS00010            None   \n",
      "3  NC_004829.2  3147     CDS  None  MGA_RS00010  MGA_RS00010  WP_011113231.1   \n",
      "4  NC_004829.2  4548    gene  None  MGA_RS00015  MGA_RS00015            None   \n",
      "\n",
      "   start strand  \n",
      "0   1892      +  \n",
      "1   1892      +  \n",
      "2   2670      +  \n",
      "3   2670      +  \n",
      "4   3162      +  \n",
      "\n",
      "### Parsing genbank file /users/lserrano/mweber/Databases/RefSeq/Release_2016_10/Genomes_extracted/GCF_000183385.1_ASM18338v1_genomic.gbff\n",
      "Nb of genome entries in Genbank file: 1\n",
      "NC_014760.1 Mycoplasma bovis PG45 clone MU clone A2, complete genome.\n",
      "Converting DNA sequence from genbank to fasta.\n",
      "Writing fasta file: /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_004829.2/NC_014760.1.fna\n",
      "len(genomeBio.features): 1924\n",
      "CDSDf:\n",
      "     chromosome   end feature  gene                id         locus_tag  \\\n",
      "0  NC_014760.1  1409    gene  None  MBOVPG45_RS00005  MBOVPG45_RS00005   \n",
      "1  NC_014760.1  1409     CDS  None  MBOVPG45_RS00005  MBOVPG45_RS00005   \n",
      "2  NC_014760.1  2652    gene  None  MBOVPG45_RS00010  MBOVPG45_RS00010   \n",
      "3  NC_014760.1  2652     CDS  None  MBOVPG45_RS00010  MBOVPG45_RS00010   \n",
      "4  NC_014760.1  2882    gene  None  MBOVPG45_RS00015  MBOVPG45_RS00015   \n",
      "\n",
      "       protein_id  start strand  \n",
      "0            None      8      +  \n",
      "1  WP_013456605.1      8      +  \n",
      "2            None   1542      +  \n",
      "3  WP_013456452.1   1542      +  \n",
      "4            None   2654      +  \n",
      "Writing CDS BED file NC_004829.2_CDS.bed\n",
      "Writing rRNA BED file NC_004829.2_rRNA.bed\n",
      "Writing tRNA BED file NC_004829.2_tRNA.bed\n",
      "Writing rRNA_tRNA BED file NC_004829.2_rRNA_tRNA.bed\n",
      "We build the bowtie2 index from the following set of fasta files:\n",
      " /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_004829.2/NC_004829.2.fna\n",
      "/users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_004829.2/NC_014760.1.fna\n",
      "/users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_013948.1/example.fasta\n",
      "### Running bowtie2-build...\n",
      "/users/lserrano/mweber/Software/bowtie2-2.2.9/bowtie2-build /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_004829.2/NC_004829.2.fna,/users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_004829.2/NC_014760.1.fna,/users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_013948.1/example.fasta NC_004829.2\n",
      "b'Settings:\n",
      "  Output files: \"NC_004829.2.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_004829.2/NC_004829.2.fna\n",
      "  /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_004829.2/NC_014760.1.fna\n",
      "  /users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_013948.1/example.fasta\n",
      "Building a SMALL index\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 504201\n",
      "Using parameters --bmax 378151 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 378151 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 6; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 1; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 252100 (target: 378150)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 8\n",
      "  Reserving size (378151) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 172435 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 172436 for bucket 1\n",
      "Getting block 2 of 8\n",
      "  Reserving size (378151) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 374225 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 374226 for bucket 2\n",
      "Getting block 3 of 8\n",
      "  Reserving size (378151) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 152520 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 152521 for bucket 3\n",
      "Getting block 4 of 8\n",
      "  Reserving size (378151) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 280245 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 280246 for bucket 4\n",
      "Getting block 5 of 8\n",
      "  Reserving size (378151) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 328997 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 328998 for bucket 5\n",
      "Getting block 6 of 8\n",
      "  Reserving size (378151) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 158853 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 158854 for bucket 6\n",
      "Getting block 7 of 8\n",
      "  Reserving size (378151) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 272405 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 272406 for bucket 7\n",
      "Getting block 8 of 8\n",
      "  Reserving size (378151) for bucket 8\n",
      "  Calculating Z arrays for bucket 8\n",
      "  Entering block accumulator loop for bucket 8:\n",
      "  bucket 8: 10%\n",
      "  bucket 8: 20%\n",
      "  bucket 8: 30%\n",
      "  bucket 8: 40%\n",
      "  bucket 8: 50%\n",
      "  bucket 8: 60%\n",
      "  bucket 8: 70%\n",
      "  bucket 8: 80%\n",
      "  bucket 8: 90%\n",
      "  bucket 8: 100%\n",
      "  Sorting block of length 277117 for bucket 8\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 277118 for bucket 8\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 700717\n",
      "fchr[G]: 1008889\n",
      "fchr[T]: 1313758\n",
      "fchr[$]: 2016804\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4866992 bytes to primary EBWT file: NC_004829.2.1.bt2\n",
      "Wrote 504208 bytes to secondary EBWT file: NC_004829.2.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 2016804\n",
      "    bwtLen: 2016805\n",
      "    sz: 504201\n",
      "    bwtSz: 504202\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 126051\n",
      "    offsSz: 504204\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 10505\n",
      "    numLines: 10505\n",
      "    ebwtTotLen: 672320\n",
      "    ebwtTotSz: 672320\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:02\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 504201\n",
      "Using parameters --bmax 378151 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 378151 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 6; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 1; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 288114 (target: 378150)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (378151) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 343375 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 343376 for bucket 1\n",
      "Getting block 2 of 7\n",
      "  Reserving size (378151) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 273058 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 273059 for bucket 2\n",
      "Getting block 3 of 7\n",
      "  Reserving size (378151) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 277405 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 277406 for bucket 3\n",
      "Getting block 4 of 7\n",
      "  Reserving size (378151) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 317501 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 317502 for bucket 4\n",
      "Getting block 5 of 7\n",
      "  Reserving size (378151) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 194472 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 194473 for bucket 5\n",
      "Getting block 6 of 7\n",
      "  Reserving size (378151) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 369618 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 369619 for bucket 6\n",
      "Getting block 7 of 7\n",
      "  Reserving size (378151) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 241369 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 241370 for bucket 7\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 700717\n",
      "fchr[G]: 1008889\n",
      "fchr[T]: 1313758\n",
      "fchr[$]: 2016804\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4866992 bytes to primary EBWT file: NC_004829.2.rev.1.bt2\n",
      "Wrote 504208 bytes to secondary EBWT file: NC_004829.2.rev.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 2016804\n",
      "    bwtLen: 2016805\n",
      "    sz: 504201\n",
      "    bwtSz: 504202\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 126051\n",
      "    offsSz: 504204\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 10505\n",
      "    numLines: 10505\n",
      "    ebwtTotLen: 672320\n",
      "    ebwtTotSz: 672320\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:02\n",
      "'\n",
      "### bowtie2-build finished\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'/users/lserrano/mweber/Translation_model/Ribosome_profiling_data/bowtie2_indexed_genome/NC_004829.2/NC_004829.2'"
      ]
     },
     "execution_count": 236,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indexedGenomePath = index_genome_files_bowtie2(genbankFileList=genbankFileList, fastaFileList=fastaFileList,\n",
    "                                             outputName=args.outputName, outputDir=args.outputDir)\n",
    "indexedGenomePath"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Appendix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import NCBI assembly summary"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I downloaded the new assembly_summary file, which has a new field I think. The problem is that we have many lines with a variable number of fields,\n",
    "```bash\n",
    "wc -l assembly_summary_refseq.txt\n",
    "86132 assembly_summary_refseq.txt\n",
    "awk -F'\\t' 'NF==21 {print NR;}' assembly_summary_refseq.txt | wc -l\n",
    "85437\n",
    "awk -F'\\t' 'NF==22 {print NR;}' assembly_summary_refseq.txt | wc -l\n",
    "694\n",
    "```\n",
    "\n",
    "Now I just changed the parsing function and I only use the first 21 columns. This does not work because the pandas parser is not able to handle varying column numbers per line. I had to correct manually the file,\n",
    "```bash\n",
    "cut -f1-21 assembly_summary_refseq.txt > assembly_summary_refseq.txt.corrected\n",
    "awk -F'\\t' 'NF==21 {print NR;}' assembly_summary_refseq.txt.corrected | wc -l\n",
    "86131\n",
    "awk -F'\\t' 'NF==22 {print NR;}' assembly_summary_refseq.txt.corrected | wc -l\n",
    "0\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nb of genomes in RefSeq database: 86130\n",
      "Nb of genomes in RefSeq database, representative or reference genomes: 5907\n",
      "Nb of genomes in RefSeq database, after filtering out contigs: 1788\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>wgs_master</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>taxid</th>\n",
       "      <th>species_taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>infraspecific_name</th>\n",
       "      <th>isolate</th>\n",
       "      <th>version_status</th>\n",
       "      <th>...</th>\n",
       "      <th>release_type</th>\n",
       "      <th>genome_rep</th>\n",
       "      <th>seq_rel_date</th>\n",
       "      <th>asm_name</th>\n",
       "      <th>submitter</th>\n",
       "      <th>gbrs_paired_asm</th>\n",
       "      <th>paired_asm_comp</th>\n",
       "      <th>ftp_path</th>\n",
       "      <th>excluded_from_refseq</th>\n",
       "      <th>compressedGenomeFilename</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_000002515.2</th>\n",
       "      <td>PRJNA12377</td>\n",
       "      <td>SAMEA3138170</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>28985</td>\n",
       "      <td>28985</td>\n",
       "      <td>Kluyveromyces lactis</td>\n",
       "      <td>strain=NRRL Y-1140</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>Major</td>\n",
       "      <td>Full</td>\n",
       "      <td>2004/07/02</td>\n",
       "      <td>ASM251v1</td>\n",
       "      <td>Genolevures Consortium</td>\n",
       "      <td>GCA_000002515.1</td>\n",
       "      <td>different</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCF_000002515.2_ASM251v1_genomic.gbff.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000002725.2</th>\n",
       "      <td>PRJNA15564</td>\n",
       "      <td>SAMEA3138173</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>347515</td>\n",
       "      <td>5664</td>\n",
       "      <td>Leishmania major strain Friedlin</td>\n",
       "      <td>strain=Friedlin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>Major</td>\n",
       "      <td>Full</td>\n",
       "      <td>2011/02/14</td>\n",
       "      <td>ASM272v2</td>\n",
       "      <td>Friedlin Consortium</td>\n",
       "      <td>GCA_000002725.2</td>\n",
       "      <td>identical</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCF_000002725.2_ASM272v2_genomic.gbff.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000002985.6</th>\n",
       "      <td>PRJNA158</td>\n",
       "      <td>SAMEA3138177</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>6239</td>\n",
       "      <td>6239</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>strain=Bristol N2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>Major</td>\n",
       "      <td>Full</td>\n",
       "      <td>2013/02/07</td>\n",
       "      <td>WBcel235</td>\n",
       "      <td>C. elegans Sequencing Consortium</td>\n",
       "      <td>GCA_000002985.3</td>\n",
       "      <td>different</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCF_000002985.6_WBcel235_genomic.gbff.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    bioproject     biosample wgs_master  \\\n",
       "assembly_accession                                        \n",
       "GCF_000002515.2     PRJNA12377  SAMEA3138170        NaN   \n",
       "GCF_000002725.2     PRJNA15564  SAMEA3138173        NaN   \n",
       "GCF_000002985.6       PRJNA158  SAMEA3138177        NaN   \n",
       "\n",
       "                          refseq_category   taxid  species_taxid  \\\n",
       "assembly_accession                                                 \n",
       "GCF_000002515.2     representative genome   28985          28985   \n",
       "GCF_000002725.2     representative genome  347515           5664   \n",
       "GCF_000002985.6          reference genome    6239           6239   \n",
       "\n",
       "                                       organism_name  infraspecific_name  \\\n",
       "assembly_accession                                                         \n",
       "GCF_000002515.2                 Kluyveromyces lactis  strain=NRRL Y-1140   \n",
       "GCF_000002725.2     Leishmania major strain Friedlin     strain=Friedlin   \n",
       "GCF_000002985.6               Caenorhabditis elegans   strain=Bristol N2   \n",
       "\n",
       "                   isolate version_status  \\\n",
       "assembly_accession                          \n",
       "GCF_000002515.2        NaN         latest   \n",
       "GCF_000002725.2        NaN         latest   \n",
       "GCF_000002985.6        NaN         latest   \n",
       "\n",
       "                                      ...                    release_type  \\\n",
       "assembly_accession                    ...                                   \n",
       "GCF_000002515.2                       ...                           Major   \n",
       "GCF_000002725.2                       ...                           Major   \n",
       "GCF_000002985.6                       ...                           Major   \n",
       "\n",
       "                   genome_rep seq_rel_date  asm_name  \\\n",
       "assembly_accession                                     \n",
       "GCF_000002515.2          Full   2004/07/02  ASM251v1   \n",
       "GCF_000002725.2          Full   2011/02/14  ASM272v2   \n",
       "GCF_000002985.6          Full   2013/02/07  WBcel235   \n",
       "\n",
       "                                           submitter  gbrs_paired_asm  \\\n",
       "assembly_accession                                                      \n",
       "GCF_000002515.2               Genolevures Consortium  GCA_000002515.1   \n",
       "GCF_000002725.2                  Friedlin Consortium  GCA_000002725.2   \n",
       "GCF_000002985.6     C. elegans Sequencing Consortium  GCA_000002985.3   \n",
       "\n",
       "                   paired_asm_comp  \\\n",
       "assembly_accession                   \n",
       "GCF_000002515.2          different   \n",
       "GCF_000002725.2          identical   \n",
       "GCF_000002985.6          different   \n",
       "\n",
       "                                                             ftp_path  \\\n",
       "assembly_accession                                                      \n",
       "GCF_000002515.2     ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...   \n",
       "GCF_000002725.2     ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...   \n",
       "GCF_000002985.6     ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...   \n",
       "\n",
       "                   excluded_from_refseq  \\\n",
       "assembly_accession                        \n",
       "GCF_000002515.2                     NaN   \n",
       "GCF_000002725.2                     NaN   \n",
       "GCF_000002985.6                     NaN   \n",
       "\n",
       "                                    compressedGenomeFilename  \n",
       "assembly_accession                                            \n",
       "GCF_000002515.2     GCF_000002515.2_ASM251v1_genomic.gbff.gz  \n",
       "GCF_000002725.2     GCF_000002725.2_ASM272v2_genomic.gbff.gz  \n",
       "GCF_000002985.6     GCF_000002985.6_WBcel235_genomic.gbff.gz  \n",
       "\n",
       "[3 rows x 21 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assemblySummaryRepDf = read_assemblySummary_file(str(refSeqPath / 'assembly_summary_refseq.txt.corrected'))\n",
    "assemblySummaryRepDf.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Search for genome assembly (`GCF_`) that contains the genome DNA sequence (`NC_`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genomesCompressedPath = refSeqPath / 'Genomes_compressed'\n",
    "genomesExtractedPath = refSeqPath / 'Genomes_extracted'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genomeSeqAccList = \\\n",
    "[\n",
    "    'NC_013948.1',\n",
    "    'NC_014760.1',\n",
    "    'NZ_CP001668.1',\n",
    "    'NC_007633.1',\n",
    "    'NC_017502.1',\n",
    "    'NC_004829.2',\n",
    "    'NC_000908.2',\n",
    "    'NC_006360.1'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "genomeFilename = None\n",
    "for fname in list(genomesExtractedPath.glob('GCF_*')):\n",
    "    if fname.is_file():\n",
    "        with fname.open() as f:\n",
    "            for line in f:\n",
    "                if 'NC_013948.1' in line:\n",
    "                    print(\"Found genome file \", fname)\n",
    "                    genomeFilename = fname\n",
    "                    break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is just extremely slow, either with python or with grep... Rather, we will query the NCBI website directly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genomeAccessionList = \\\n",
    "[\n",
    "    'GCF_000089865.1',\n",
    "    'GCF_000183385.1',\n",
    "    'GCF_000023685.1',\n",
    "    'GCF_000012765.1',\n",
    "    'GCF_000025365.1',\n",
    "    'GCF_000092585.1',\n",
    "    'GCF_000027325.1',\n",
    "    'GCF_000008405.1'\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In general this is a bad idea, because it is not straightforward to find which genome assembly contains the genome DNA sequence accession. It is better to start the script directly with the genbank file of the genome assembly or the genome DNA sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genomeAcc = genomeSeqAccList[0]\n",
    "genomePath = rnaSeqDataPath / 'bowtie2_indexed_genome' / genomeAcc\n",
    "genbankFilename = str(genomePath / (genomeAcc + '.gb')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
